openmicroscopy/bioformats
# ome-xml: Use language-specific container types for maps

This makes some (non-breaking) model changes which result in some breaking model (but not metadata) API changes.  The overall aim is to allow the C++ code generation to function; the existing approach was fundamentally incompatible with metadata conversion in terms of having a usable concrete map container instance which could be safely used with get/set methods on the metadatastore to transfer metadata between stores.  It can also be considered to be fixing some historical design warts with the Map model and implementation which clean it up and make the code generator, templates and resulting API saner, and should perhaps have been done at the time we originally introduced these features.

The type used with the MetadataStore can possibly be reverted if this is unacceptable.  The model changes are needed, however.

Also note that we might need to make the corresponding changes to the 2015 schema to get C++ to work with it; that wouldn't be a breaking change, but it might be something we want to think carefully about before doing.

See also: https://github.com/openmicroscopy/openmicroscopy/pull/4671 for the corresponding OMERO changes.

--------

Summary of changes

model:
  - `Map` is a complexType, not an element.  This removes the generation of the useless `Map` model object.  This does not change the schema; it was never used as a concrete element.  This is needed for code generation sanity (the generated inheritance hierarchy).
  - Drop `MapPairs`; this is also non-breaking since it was also never used directly in OME-XML.
  - `MapAnnotation` `Value`, `ImagingEnvironment` `Map` and `GenericExcitationSource` `Map` are now implemented in terms of the `Map` complexType.
  The above changes do not change the validation of the OME-XML in any shape or form.

xsd-fu:
  - Add Map support to the C++ templates; it's currently disabled when using the 2013-06 schema.
  - At the model object level, redundant `getPairs()` calls are eliminated entirely.  The model object template handles `List<MapPair` serialisation directly.

cpp:
  - Add `OrderedMultimap` model primitive and tests to validate duplicate keys and key/value pairs, and preservation of insertion order.
  - Drop (unused) `MapPairs` model object.

java:
  - Simplify xsd-fu template usage; List<MapPair> is a model primitive.
  - Drop `MapPairs`; this type is now redundant.
  - Unit tests check that the same key may be used multiple times and that insertion order is preserved.  They also verify that the same key/value pair may be resused.

Source changes: https://ci.openmicroscopy.org/job/BIOFORMATS-DEV-breaking-generated-sources/219/

--------

Testing:

- Check generated source code differences
    - https://ci.openmicroscopy.org/job/BIOFORMATS-DEV-breaking-generated-sources/lastSuccessfulBuild/artifact/ome-xml-metadata-java.diff/*view*/
    - https://ci.openmicroscopy.org/job/BIOFORMATS-DEV-breaking-generated-sources/lastSuccessfulBuild/artifact/ome-xml-model-java.diff/*view*/
- Check kyro can serialise readers using the different types of Map annotation
- Check OMERO can import files with map annotations
- You'll see that there are no C++ model changes in this PR; these will be enabled with the 2016-DEV0/2015 model in a followup PR; what's here has been built and tested on my `model-units-cpp3` branch (from which this PR was factored out).  C++ testing is limited to checking the jobs remain green.
